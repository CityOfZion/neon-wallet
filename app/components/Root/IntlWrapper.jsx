// @flow
import React from 'react'
import { compose } from 'recompose'
import '@formatjs/intl-pluralrules/polyfill'
import '@formatjs/intl-pluralrules/polyfill-locales'
import { IntlProvider } from 'react-intl'

import translations from '../../translations'
import { LANGUAGES, DEFAULT_LANGUAGE } from '../../core/constants'
import { useSettingsContext } from '../../context/settings/SettingsContext'

const {
  english,
  korean,
  french,
  vietnamese,
  chinese,
  german,
  italian,
  portuguese,
  turkish,
  arabic,
  russian,
  dutch,
} = translations

type IntlWrapperProps = {
  children: any,
}

const tranlationsMappings = {
  [LANGUAGES.ENGLISH.value]: english,
  [LANGUAGES.KOREAN.value]: korean,
  [LANGUAGES.FRENCH.value]: french,
  [LANGUAGES.VIETNAMESE.value]: vietnamese,
  [LANGUAGES.GERMAN.value]: german,
  [LANGUAGES.CHINESE.value]: chinese,
  [LANGUAGES.ITALIAN.value]: italian,
  [LANGUAGES.PORTUGUESE.value]: portuguese,
  [LANGUAGES.TURKISH.value]: turkish,
  [LANGUAGES.ARABIC.value]: arabic,
  [LANGUAGES.RUSSIAN.value]: russian,
  [LANGUAGES.DUTCH.value]: dutch,
}

function IntlWrapper(props: IntlWrapperProps) {
  const { children } = props
  const { settings } = useSettingsContext()

  const language = settings
    ? settings.language || DEFAULT_LANGUAGE
    : DEFAULT_LANGUAGE

  return (
    <IntlProvider
      locale="en"
      messages={tranlationsMappings[language] || english}
    >
      {children}
    </IntlProvider>
  )
}

export default IntlWrapper
