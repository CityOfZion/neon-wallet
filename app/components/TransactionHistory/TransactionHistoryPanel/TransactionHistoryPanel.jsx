// @flow
import React from 'react'
import classNames from 'classnames'
import { intersectionBy } from 'lodash-es'

import Transactions from './Transactions'
import Panel from '../../Panel'
import { pruneConfirmedOrStaleTransaction } from '../../../actions/pendingTransactionActions'
import styles from './TransactionHistoryPanel.scss'

type Props = {
  className: ?string,
  transactions: Array<Object>,
  handleFetchAdditionalTxData: () => void,
  handleGetPendingTransactionInfo: () => void,
  handleRefreshTxData: () => void,
  pendingTransactions: Array<Object>,
  address: string,
  showSuccessNotification: ({ message: string }) => void,
}

const REFRESH_INTERVAL_MS = 30000

export default class TransactionHistory extends React.Component<Props> {
  transactionDataInterval: IntervalID

  static defaultProps = {
    transactions: [],
    pendingTransactions: [],
  }

  componentDidMount() {
    this.addPolling()
  }

  componentWillUnmount() {
    this.removePolling()
  }

  render() {
    const { className, transactions } = this.props
    const filteredPendingTransactions = this.pruneConfirmedTransactionsFromPending()
    this.pruneReturnedTransactionsFromStorage()
    return (
      <Panel
        className={classNames(styles.transactionHistoryPanel, className)}
        onScroll={this.handleScroll}
      >
        <Transactions
          className={styles.transactions}
          transactions={transactions}
          pendingTransactions={filteredPendingTransactions || []}
        />
      </Panel>
    )
  }

  addPolling = () => {
    const { showSuccessNotification } = this.props
    this.transactionDataInterval = setInterval(async () => {
      await this.props.handleGetPendingTransactionInfo()
      this.props.handleRefreshTxData()
      showSuccessNotification({
        message: 'Received latest transaction information.',
      })
    }, REFRESH_INTERVAL_MS)
  }

  removePolling = () => {
    if (this.transactionDataInterval) {
      clearInterval(this.transactionDataInterval)
    }
  }

  pruneConfirmedTransactionsFromPending() {
    const { transactions, pendingTransactions } = this.props
    const confirmed = transactions.map(tx => tx.txid)
    return pendingTransactions.reduce((accum, currVal) => {
      if (confirmed.find(tx => tx === currVal.txid)) return accum
      accum.push(currVal)
      return accum
    }, [])
  }

  async pruneReturnedTransactionsFromStorage() {
    const {
      transactions,
      pendingTransactions,
      address,
      handleGetPendingTransactionInfo,
    } = this.props
    const toBePurged = intersectionBy(transactions, pendingTransactions, 'txid')
    // eslint-disable-next-line
    for (const transaction of toBePurged) {
      // eslint-disable-next-line
      await pruneConfirmedOrStaleTransaction(address, transaction.txid)
      handleGetPendingTransactionInfo()
    }
  }

  handleScroll = (e: SyntheticInputEvent<EventTarget>) => {
    const { handleFetchAdditionalTxData } = this.props
    const bottom =
      e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight
    if (bottom) {
      handleFetchAdditionalTxData()
    }
  }
}
