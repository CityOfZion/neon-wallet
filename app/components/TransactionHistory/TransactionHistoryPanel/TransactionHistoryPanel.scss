@import '../../../styles/variables';

.transactionInfo {
  display: flex;
  flex-direction: column;
  text-align: center;
  position: relative;
  height: 100%;

  .transactions {
    flex: 1 1 auto;
    overflow-y: auto;
  }
}

.updateLoader {
  right: 0;
  left: auto;
  top: -5px;
}

.oddNumberedRow {
  background: #f8f8f8;
}

.transactions {
  margin: -24px;
}

.loadMoreButton {
  height: 40px;
}

.transactionHistoryPanel {
  margin-top: -12px;
}

.buttonLoadingIndicator {
  position: relative;
  width: 15px;
  height: 15px;
  opacity: 0.4;
  margin: auto;
  left: -10px;
  top: auto;
  transform: none;
}
