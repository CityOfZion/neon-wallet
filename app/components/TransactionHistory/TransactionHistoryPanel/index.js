// @flow
import { compose } from 'recompose'
import { withData, withActions, withCall } from 'spunky'
import { connect } from 'react-redux'

import { bindActionCreators } from 'redux'
import TransactionHistoryPanel from './TransactionHistoryPanel'
import transactionHistoryActions from '../../../actions/transactionHistoryActions'
import { getPendingTransactionInfo } from '../../../actions/pendingTransactionActions'
import withProgressPanel from '../../../hocs/withProgressPanel'
import withAuthData from '../../../hocs/withAuthData'
import withNetworkData from '../../../hocs/withNetworkData'
import withLoadingProp from '../../../hocs/withLoadingProp'
import {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification,
} from '../../../modules/notifications'

const mapTransactionsDataToProps = data => ({
  transactions: data?.entries ?? [],
  count: data?.count ?? 0,
})

const mapAccountActionsToProps = (actions, { net, address }) => ({
  handleFetchAdditionalTxData: () =>
    actions.call({
      net,
      address,
      shouldIncrementPagination: true,
    }),
  handleRefreshTxData: () =>
    actions.call({
      net,
      address,
      shouldIncrementPagination: false,
    }),
})

const mapPendingTransactionActionsToProps = (actions, { net, address }) => ({
  handleGetPendingTransactionInfo: () =>
    actions.call({
      net,
      address,
    }),
})

const mapPendingTransactionInfoToProps = pendingTransactions => ({
  pendingTransactions: pendingTransactions || [],
})

const actionCreators = {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification,
}

const mapDispatchToProps = dispatch =>
  bindActionCreators(actionCreators, dispatch)

export default compose(
  connect(
    null,
    mapDispatchToProps,
  ),
  withAuthData(),
  withNetworkData(),
  // withProgressPanel(transactionHistoryActions, {
  //   title: '',
  // }),
  withActions(transactionHistoryActions, mapAccountActionsToProps),
  withActions(getPendingTransactionInfo, mapPendingTransactionActionsToProps),
  withCall(getPendingTransactionInfo),
  withData(getPendingTransactionInfo, mapPendingTransactionInfoToProps),
  withLoadingProp(transactionHistoryActions),
  withData(transactionHistoryActions, mapTransactionsDataToProps),
)(TransactionHistoryPanel)
