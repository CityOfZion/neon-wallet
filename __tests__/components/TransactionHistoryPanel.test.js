import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import { merge } from 'lodash-es'
import { progressValues } from 'spunky'

import IntlWrapper from '../../app/components/Root/IntlWrapper'
import TransactionHistoryPanel from '../../app/components/TransactionHistory/TransactionHistoryPanel'
import {
  MAIN_NETWORK_ID,
  EXPLORERS,
  DEFAULT_LANGUAGE,
} from '../../app/core/constants'

const { LOADED } = progressValues

const initialState = {
  spunky: {
    auth: {
      batch: false,
      progress: LOADED,
      data: {
        address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh',
        wif: 'L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou',
      },
    },
    network: {
      batch: false,
      progress: LOADED,
      loadedCount: 1,
      data: MAIN_NETWORK_ID,
    },
    settings: {
      batch: false,
      progress: LOADED,
      loadedCount: 1,
      data: {
        blockExplorer: EXPLORERS.NEO_TRACKER,
        language: DEFAULT_LANGUAGE,
      },
    },
    transactionHistory: {
      batch: false,
      progress: LOADED,
      loadedCount: 1,
      data: { transactions: [] },
    },
  },
}

const transactions = [
  {
    to: 'AXmzKD3dvj7dPUQKkBeNZmRDYF6AhrwrtQ',
    from: 'claim',
    txid: '6ab0391ee237c7b2516f13d47f09a30844da64fefefc006877b40542d34a705b',
    time: 1537278057,
    amount: '0.11988459',
    asset: { symbol: 'GAS' },
    label: 'Gas Claim',
    type: 'CLAIM',
    id: '_ymelbt8mb',
  },
  {
    to: 'AXmzKD3dvj7dPUQKkBeNZmRDYF6AhrwrtQ',
    from: 'AXmzKD3dvj7dPUQKkBeNZmRDYF6AhrwrtQ',
    txid: '1ec4446bac87fca59abff2ccf96283547a5ebec64b7c1aba7a31fb91b93c9b63',
    time: 1537234593,
    amount: '0',
    asset: {
      symbol: 'MCT',
    },
    label: 'MCT',
    type: 'RECEIVE',
    id: '_r3mihxg36',
  },
]

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<TransactionHistoryPanel store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <IntlWrapper>
          <TransactionHistoryPanel />
        </IntlWrapper>
      </Provider>,
    )
  }

  return {
    store,
    wrapper,
  }
}

describe('TransactionHistoryPanel', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('correctly renders no transaction history', () => {
    const { wrapper } = setup(initialState, false)

    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(0)
  })

  test('correctly renders with NEO and GAS transaction history', () => {
    const transactionState = merge({}, initialState, {
      spunky: {
        transactionHistory: { data: { entries: transactions, count: 2 } },
      },
    })
    const { wrapper } = setup(transactionState, false)
    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(2)

    expect(
      transactionList
        .childAt(0)
        .find('.txAmountContainer')
        .text(),
    ).toEqual('0.11988459')

    expect(
      transactionList
        .childAt(1)
        .find('.txAmountContainer')
        .text(),
    ).toEqual('0')

    expect(
      transactionList
        .childAt(1)
        .find('.txLabelContainer')
        .text(),
    ).toEqual('MCT')

    expect(
      transactionList
        .childAt(0)
        .find('.txLabelContainer')
        .text(),
    ).toEqual('Gas Claim')

    expect(
      transactionList
        .childAt(0)
        .find('.txToContainer')
        .text(),
    ).toEqual(transactions[0].to)

    expect(
      transactionList
        .childAt(1)
        .find('.txToContainer')
        .text(),
    ).toEqual(transactions[1].to)
  })
})
