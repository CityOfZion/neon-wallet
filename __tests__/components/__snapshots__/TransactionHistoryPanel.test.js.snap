// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHistoryPanel renders without crashing 1`] = `
<Connect(withData(Connect(withData(withProps(Connect(withActions(Connect(withActions(Connect(withCall(Connect(withData(Connect(withProgress(withProps(Connect(withData(TransactionHistory))))))))))))))))))
  showErrorNotification={[Function]}
  showInfoNotification={[Function]}
  showSuccessNotification={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
